### PR TITLE
Implement Compact codec for alloy_primitives::Log

### PR DESCRIPTION
**Pull request by ghost**
Ref https://github.com/paradigmxyz/reth/issues/7050

### CODE REVIEW

File: crates/storage/codecs/src/lib.rs [hunk 1]
```diff
@@ -13,9 +13,55 @@
 
 pub use reth_codecs_derive::*;
 
-use alloy_primitives::{Address, Bloom, Bytes, B256, B512, U256};
+use alloy_primitives::{Address, Bloom, Bytes, Log, LogData, B256, B512, U256};
 use bytes::Buf;
 
+impl Compact for LogData {
```
Comment (mattsse) [inline crates/storage/codecs/src/lib.rs:19]: this should be moved to a separate file

**Review by mattsse — CHANGES_REQUESTED** [review]
this is a great start,

it is absolutely crucial that we have the same manual impl as currently on on main, so what we need is a bunch of test vectors of encoded reth-primitive types data.

It would also be helpful if you could attach the cargo expand of the reth-primitive types to the issue

**Comment by 0xqd** [comment]
@mattsse cargo expand file on Log

```rust
mod log {
    use crate::{Address, Bloom, Bytes, B256};
    use alloy_primitives::Log as AlloyLog;
    use alloy_rlp::{RlpDecodable, RlpEncodable};
    use reth_codecs::{main_codec, Compact};
    /// Ethereum Log
    pub struct Log {
        /// Contract that emitted this log.
        pub address: Address,
        /// Topics of the log. The number of logs depend on what `LOG` opcode is used.
        pub topics: Vec<B256>,
        /// Arbitrary length data.
        pub data: Bytes,
    }
    #[automatically_derived]
    impl ::core::clone::Clone for Log {
        #[inline]
        fn clone(&self) -> Log {
            Log {
                address: ::core::clone::Clone::clone(&self.address),
                topics: ::core::clone::Clone::clone(&self.topics),
                data: ::core::clone::Clone::clone(&self.data),
            }
        }
    }
    #[automatically_derived]
    impl ::core::fmt::Debug for Log {
        #[inline]
        fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
            ::core::fmt::Formatter::debug_struct_field3_finish(
                f,
                "Log",
                "address",
                &self.address,
                "topics",
                &self.topics,
                "data",
                &&self.data,
            )
        }
    }
    #[automatically_derived]
    impl ::core::marker::StructuralPartialEq for Log {}
    #[automatically_derived]
    impl ::core::cmp::PartialEq for Log {
        #[inline]
        fn eq(&self, other: &Log) -> bool {
            self.address == other.address && self.topics == other.topics
                && self.data == other.data
        }
    }
    #[automatically_derived]
    impl ::core::cmp::Eq for Log {
        #[inline]
        #[doc(hidden)]
        #[coverage(off)]
        fn assert_receiver_is_total_eq(&self) -> () {
            let _: ::core::cmp::AssertParamIsEq<Address>;
            let _: ::core::cmp::AssertParamIsEq<Vec<B256>>;
            let _: ::core::cmp::AssertParamIsEq<Bytes>;
        }
    }
    const _: () = {
        extern crate alloy_rlp;
        impl alloy_rlp::Decodable for Log {
            #[inline]
            fn decode(b: &mut &[u8]) -> alloy_rlp::Result<Self> {
                let alloy_rlp::Header { list, payload_length } = alloy_rlp::Header::decode(
                    b,
                )?;
                if !list {
                    return Err(alloy_rlp::Error::UnexpectedString);
                }
                let started_len = b.len();
                if started_len < payload_length {
                    return Err(alloy_rlp::DecodeError::InputTooShort);
                }
                let this = Self {
                    address: alloy_rlp::Decodable::decode(b)?,
                    topics: alloy_rlp::Decodable::decode(b)?,
                    data: alloy_rlp::Decodable::decode(b)?,
                };
                let consumed = started_len - b.len();
                if consumed != payload_length {
                    return Err(alloy_rlp::Error::ListLengthMismatch {
                        expected: payload_length,
                        got: consumed,
                    });
                }
                Ok(this)
            }
        }
    };
    const _: () = {
        extern crate alloy_rlp;
        impl alloy_rlp::Encodable for Log {
            #[inline]
            fn length(&self) -> usize {
                let payload_length = self._alloy_rlp_payload_length();
                payload_length + alloy_rlp::length_of_length(payload_length)
            }
            #[inline]
            fn encode(&self, out: &mut dyn alloy_rlp::BufMut) {
                alloy_rlp::Header {
                    list: true,
                    payload_length: self._alloy_rlp_payload_length(),
                }
                    .encode(out);
                alloy_rlp::Encodable::encode(&self.address, out);
                alloy_rlp::Encodable::encode(&self.topics, out);
                alloy_rlp::Encodable::encode(&self.data, out);
            }
        }
        impl Log {
            #[allow(unused_parens)]
            #[inline]
            fn _alloy_rlp_payload_length(&self) -> usize {
                0usize + alloy_rlp::Encodable::length(&self.address)
                    + alloy_rlp::Encodable::length(&self.topics)
                    + alloy_rlp::Encodable::length(&self.data)
            }
        }
    };
    #[automatically_derived]
    impl ::core::default::Default for Log {
        #[inline]
        fn default() -> Log {
            Log {
                address: ::core::default::Default::default(),
                topics: ::core::default::Default::default(),
                data: ::core::default::Default::default(),
            }
        }
    }
    /// Placeholder struct for when there is no need for a fieldset. Doesn't actually write or read any data.
    pub struct LogFlags {}
    #[automatically_derived]
    impl ::core::fmt::Debug for LogFlags {
        #[inline]
        fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
            ::core::fmt::Formatter::write_str(f, "LogFlags")
        }
    }
    #[automatically_derived]
    impl ::core::default::Default for LogFlags {
        #[inline]
        fn default() -> LogFlags {
            LogFlags {}
        }
    }
    impl LogFlags {
        /// Placeholder: does not read any value.
        pub fn from(mut buf: &[u8]) -> (Self, &[u8]) {
            (LogFlags::default(), buf)
        }
        /// Placeholder: returns an empty array.
        pub fn into_bytes(self) -> [u8; 0] {
            []
        }
    }
    impl Compact for Log {
        fn to_compact<B>(self, buf: &mut B) -> usize
        where
            B: bytes::BufMut + AsMut<[u8]>,
        {
            let mut flags = LogFlags::default();
            let mut total_length = 0;
            let mut buffer = bytes::BytesMut::new();
            let address_len = self.address.to_compact(&mut buffer);
            let topics_len = self.topics.specialized_to_compact(&mut buffer);
            let data_len = self.data.to_compact(&mut buffer);
            let flags = flags.into_bytes();
            total_length += flags.len() + buffer.len();
            buf.put_slice(&flags);
            buf.put(buffer);
            total_length
        }
        fn from_compact(mut buf: &[u8], len: usize) -> (Self, &[u8]) {
            let (flags, mut buf) = LogFlags::from(buf);
            let (address, new_buf) = Address::from_compact(buf, buf.len());
            buf = new_buf;
            let (topics, new_buf) = Vec::specialized_from_compact(buf, buf.len());
            buf = new_buf;
            let mut data = Bytes::new();
            (data, buf) = Bytes::from_compact(buf, buf.len() as usize);
            let obj = Log {
                address: address,
                topics: topics,
                data: data,
            };
            (obj, buf)
        }
    }
    #[doc(hidden)]
    #[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
    const _: () = {
        #[allow(unused_extern_crates, clippy::useless_attribute)]
        extern crate serde as _serde;
        #[automatically_derived]
        impl _serde::Serialize for Log {
            fn serialize<__S>(
                &self,
                __serializer: __S,
            ) -> _serde::__private::Result<__S::Ok, __S::Error>
            where
                __S: _serde::Serializer,
            {
                let mut __serde_state = _serde::Serializer::serialize_struct(
                    __serializer,
                    "Log",
                    false as usize + 1 + 1 + 1,
                )?;
                _serde::ser::SerializeStruct::serialize_field(
                    &mut __serde_state,
                    "address",
                    &self.address,
                )?;
                _serde::ser::SerializeStruct::serialize_field(
                    &mut __serde_state,
                    "topics",
                    &self.topics,
                )?;
                _serde::ser::SerializeStruct::serialize_field(
                    &mut __serde_state,
                    "data",
                    &self.data,
                )?;
                _serde::ser::SerializeStruct::end(__serde_state)
            }
        }
    };
    #[doc(hidden)]
    #[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
    const _: () = {
        #[allow(unused_extern_crates, clippy::useless_attribute)]
        extern crate serde as _serde;
        #[automatically_derived]
        impl<'de> _serde::Deserialize<'de> for Log {
            fn deserialize<__D>(
                __deserializer: __D,
            ) -> _serde::__private::Result<Self, __D::Error>
            where
                __D: _serde::Deserializer<'de>,
            {
                #[allow(non_camel_case_types)]
                #[doc(hidden)]
                enum __Field {
                    __field0,
                    __field1,
                    __field2,
                    __ignore,
                }
                #[doc(hidden)]
                struct __FieldVisitor;
                impl<'de> _serde::de::Visitor<'de> for __FieldVisitor {
                    type Value = __Field;
                    fn expecting(
                        &self,
                        __formatter: &mut _serde::__private::Formatter,
                    ) -> _serde::__private::fmt::Result {
                        _serde::__private::Formatter::write_str(
                            __formatter,
                            "field identifier",
                        )
                    }
                    fn visit_u64<__E>(
                        self,
                        __value: u64,
                    ) -> _serde::__private::Result<Self::Value, __E>
                    where
                        __E: _serde::de::Error,
                    {
                        match __value {
                            0u64 => _serde::__private::Ok(__Field::__field0),
                            1u64 => _serde::__private::Ok(__Field::__field1),
                            2u64 => _serde::__private::Ok(__Field::__field2),
                            _ => _serde::__private::Ok(__Field::__ignore),
                        }
                    }
                    fn visit_str<__E>(
                        self,
                        __value: &str,
                    ) -> _serde::__private::Result<Self::Value, __E>
                    where
                        __E: _serde::de::Error,
                    {
                        match __value {
                            "address" => _serde::__private::Ok(__Field::__field0),
                            "topics" => _serde::__private::Ok(__Field::__field1),
                            "data" => _serde::__private::Ok(__Field::__field2),
                            _ => _serde::__private::Ok(__Field::__ignore),
                        }
                    }
                    fn visit_bytes<__E>(
                        self,
                        __value: &[u8],
                    ) -> _serde::__private::Result<Self::Value, __E>
                    where
                        __E: _serde::de::Error,
                    {
                        match __value {
                            b"address" => _serde::__private::Ok(__Field::__field0),
                            b"topics" => _serde::__private::Ok(__Field::__field1),
                            b"data" => _serde::__private::Ok(__Field::__field2),
                            _ => _serde::__private::Ok(__Field::__ignore),
                        }
                    }
                }
                impl<'de> _serde::Deserialize<'de> for __Field {
                    #[inline]
                    fn deserialize<__D>(
                        __deserializer: __D,
                    ) -> _serde::__private::Result<Self, __D::Error>
                    where
                        __D: _serde::Deserializer<'de>,
                    {
                        _serde::Deserializer::deserialize_identifier(
                            __deserializer,
                            __FieldVisitor,
                        )
                    }
                }
                #[doc(hidden)]
                struct __Visitor<'de> {
                    marker: _serde::__private::PhantomData<Log>,
                    lifetime: _serde::__private::PhantomData<&'de ()>,
                }
                impl<'de> _serde::de::Visitor<'de> for __Visitor<'de> {
                    type Value = Log;
                    fn expecting(
                        &self,
                        __formatter: &mut _serde::__private::Formatter,
                    ) -> _serde::__private::fmt::Result {
                        _serde::__private::Formatter::write_str(
                            __formatter,
                            "struct Log",
                        )
                    }
                    #[inline]
                    fn visit_seq<__A>(
                        self,
                        mut __seq: __A,
                    ) -> _serde::__private::Result<Self::Value, __A::Error>
                    where
                        __A: _serde::de::SeqAccess<'de>,
                    {
                        let __field0 = match _serde::de::SeqAccess::next_element::<
                            Address,
                        >(&mut __seq)? {
                            _serde::__private::Some(__value) => __value,
                            _serde::__private::None => {
                                return _serde::__private::Err(
                                    _serde::de::Error::invalid_length(
                                        0usize,
                                        &"struct Log with 3 elements",
                                    ),
                                );
                            }
                        };
                        let __field1 = match _serde::de::SeqAccess::next_element::<
                            Vec<B256>,
                        >(&mut __seq)? {
                            _serde::__private::Some(__value) => __value,
                            _serde::__private::None => {
                                return _serde::__private::Err(
                                    _serde::de::Error::invalid_length(
                                        1usize,
                                        &"struct Log with 3 elements",
                                    ),
                                );
                            }
                        };
                        let __field2 = match _serde::de::SeqAccess::next_element::<
                            Bytes,
                        >(&mut __seq)? {
                            _serde::__private::Some(__value) => __value,
                            _serde::__private::None => {
                                return _serde::__private::Err(
                                    _serde::de::Error::invalid_length(
                                        2usize,
                                        &"struct Log with 3 elements",
                                    ),
                                );
                            }
                        };
                        _serde::__private::Ok(Log {
                            address: __field0,
                            topics: __field1,
                            data: __field2,
                        })
                    }
                    #[inline]
                    fn visit_map<__A>(
                        self,
                        mut __map: __A,
                    ) -> _serde::__private::Result<Self::Value, __A::Error>
                    where
                        __A: _serde::de::MapAccess<'de>,
                    {
                        let mut __field0: _serde::__private::Option<Address> = _serde::__private::None;
                        let mut __field1: _serde::__private::Option<Vec<B256>> = _serde::__private::None;
                        let mut __field2: _serde::__private::Option<Bytes> = _serde::__private::None;
                        while let _serde::__private::Some(__key) = _serde::de::MapAccess::next_key::<
                            __Field,
                        >(&mut __map)? {
                            match __key {
                                __Field::__field0 => {
                                    if _serde::__private::Option::is_some(&__field0) {
                                        return _serde::__private::Err(
                                            <__A::Error as _serde::de::Error>::duplicate_field(
                                                "address",
                                            ),
                                        );
                                    }
                                    __field0 = _serde::__private::Some(
                                        _serde::de::MapAccess::next_value::<Address>(&mut __map)?,
                                    );
                                }
                                __Field::__field1 => {
                                    if _serde::__private::Option::is_some(&__field1) {
                                        return _serde::__private::Err(
                                            <__A::Error as _serde::de::Error>::duplicate_field("topics"),
                                        );
                                    }
                                    __field1 = _serde::__private::Some(
                                        _serde::de::MapAccess::next_value::<Vec<B256>>(&mut __map)?,
                                    );
                                }
                                __Field::__field2 => {
                                    if _serde::__private::Option::is_some(&__field2) {
                                        return _serde::__private::Err(
                                            <__A::Error as _serde::de::Error>::duplicate_field("data"),
                                        );
                                    }
                                    __field2 = _serde::__private::Some(
                                        _serde::de::MapAccess::next_value::<Bytes>(&mut __map)?,
                                    );
                                }
                                _ => {
                                    let _ = _serde::de::MapAccess::next_value::<
                                        _serde::de::IgnoredAny,
                                    >(&mut __map)?;
                                }
                            }
                        }
                        let __field0 = match __field0 {
                            _serde::__private::Some(__field0) => __field0,
                            _serde::__private::None => {
                                _serde::__private::de::missing_field("address")?
                            }
                        };
                        let __field1 = match __field1 {
                            _serde::__private::Some(__field1) => __field1,
                            _serde::__private::None => {
                                _serde::__private::de::missing_field("topics")?
                            }
                        };
                        let __field2 = match __field2 {
                            _serde::__private::Some(__field2) => __field2,
                            _serde::__private::None => {
                                _serde::__private::de::missing_field("data")?
                            }
                        };
                        _serde::__private::Ok(Log {
                            address: __field0,
                            topics: __field1,
                            data: __field2,
                        })
                    }
                }
                #[doc(hidden)]
                const FIELDS: &'static [&'static str] = &["address", "topics", "data"];
                _serde::Deserializer::deserialize_struct(
                    __deserializer,
                    "Log",
                    FIELDS,
                    __Visitor {
                        marker: _serde::__private::PhantomData::<Log>,
                        lifetime: _serde::__private::PhantomData,
                    },
                )
            }
        }
    };
    impl From<AlloyLog> for Log {
        fn from(mut log: AlloyLog) -> Self {
            Self {
                address: log.address,
                topics: std::mem::take(log.data.topics_mut_unchecked()),
                data: log.data.data,
            }
        }
    }
    impl From<Log> for AlloyLog {
        fn from(log: Log) -> AlloyLog {
            AlloyLog::new_unchecked(log.address, log.topics, log.data)
        }
    }
    /// Calculate receipt logs bloom.
    pub fn logs_bloom<'a, It>(logs: It) -> Bloom
    where
        It: IntoIterator<Item = &'a Log>,
    {
        let mut bloom = Bloom::ZERO;
        for log in logs {
            bloom.m3_2048(log.address.as_slice());
            for topic in &log.topics {
                bloom.m3_2048(topic.as_slice());
            }
        }
        bloom
    }
}

```

**Comment by mattsse** [comment]
this is sufficient:

https://gist.github.com/0xqd/fddc09cdddce9a244bf2ee8d79c4d573#file-log-rs-L161-L192

what we need is, generate random instances of reth-primitives::Log, store rlp/json encoded and store compat encoded, then we use these as testvectors for this pr

**Comment by 0xqd** [comment]
yeah, I think we will make it work with compact first, there is other codec implemented behind different gate:  `scale, postcard`, ...

**Comment by mattsse** [comment]
hmm not sure we need this

**Comment by 0xqd** [comment]
@mattsse for the RlpEncodable and Decodable we will enable it in the alloy_primitives::Log right? It's not possible to implement it here ( rust orphan's rule ).
Btw, after exploring this, I think we can have a better design by create a wrapper of type, something like this, it prevents deduplication and we can reduse all the macro just fine.

```rust
#[main_codec(rlp)]
#[derive(Clone, Debug, PartialEq, Eq, RlpDecodable, RlpEncodable, Default)]
pub struct LogWrapper (pub Log);

impl From<AlloyLog> for LogWrapper {
    fn from(log: AlloyLog) -> Self {
        Self(Log::from(log))
    }
}

impl From<LogWrapper> for AlloyLog {
    fn from(wrapper: LogWrapper) -> Self {
        AlloyLog::from(wrapper.0)
    }
}
```

And we have a minor thing regarding reth_primitives::Log expose topics directly, and alloy::Log exposes topics, via method in LogData. And we are good. wdyt?

File: crates/storage/codecs/src/log.rs
Comment (mattsse) [inline crates/storage/codecs/src/log.rs:1]: this should be under alloy/log.rs

**Comment by mattsse** [comment]
> @mattsse for the RlpEncodable and Decodable we will enable it in the alloy_primitives::Log right? It's not possible to implement it here ( rust orphan's rule ). Btw, after exploring this, I think we can have a better design by create a wrapper of type, something like this, it prevents deduplication and we can reuse all the macro just fine. Reimplement what macro already does also creates a leaky implementation.
> 
> ```rust
> #[main_codec(rlp)]
> #[derive(Clone, Debug, PartialEq, Eq, RlpDecodable, RlpEncodable, Default)]
> pub struct LogWrapper (pub Log);
> 
> impl From<AlloyLog> for LogWrapper {
>     fn from(log: AlloyLog) -> Self {
>         Self(Log::from(log))
>     }
> }
> 
> impl From<LogWrapper> for AlloyLog {
>     fn from(wrapper: LogWrapper) -> Self {
>         AlloyLog::from(wrapper.0)
>     }
> }
> ```
> 
> And we have a minor thing regarding reth_primitives::Log expose topics directly, and alloy::Log exposes topics, via method in LogData. And we are good. wdyt?

the entire point of this is deduplicating types, why do we need a wrapper?

**Review by mattsse — COMMENTED** [review]
cool, 
this looks identical to cargo expand, great!

now we only need so good test vectors

we can limit this pr to trait impl only, and replace the type seperately

File: crates/storage/codecs/src/alloy/log.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+use bytes::BufMut;
+
+/// Implement `Compact` for `LogData` and `Log`.
+impl Compact for LogData {
+    fn to_compact<B>(self, buf: &mut B) -> usize
+    where
+        B: BufMut + AsMut<[u8]>,
+    {
+        let mut buffer = bytes::BytesMut::new();
+        self.topics().to_vec().specialized_to_compact(&mut buffer);
```
Comment (mattsse) [inline crates/storage/codecs/src/alloy/log.rs:12]: I believe you can use `.split` here and avoid the to_vec

**Comment by 0xqd** [comment]
The current implementation of `reth_primitives::Log` is nearly a copy paste of `alloy_primitives:Log`, and the same for `revm_primitives::Log`. We have 2 options for pros and cons:

**Option 1: Thin Wrapper (LogWrapper)**

```rust
#[main_codec(rlp)]
#[derive(Clone, Debug, PartialEq, Eq, RlpDecodable, RlpEncodable, Default)]
pub struct LogWrapper (pub Log);
```

Pros:
Flexibility: You can add methods specific to reth or revm without modifying the shared alloy_primitives::Log. This allows for customization and independent development for each crate.
Decoupling: The crates using LogWrapper are less tightly coupled to alloy_primitives, potentially making future updates or changes easier to manage.

Cons:
Overhead: A thin wrapper introduces a small amount of overhead due to the additional type layer.
Type Conversions: Conversions between Log and LogWrapper might be necessary in some situations, adding complexity.

**Option 2: Shared Type (alloy_primitives::Log)**

Pros:
Simplicity: No wrapper is needed, reducing code complexity and potential overhead.
Direct Usage: Crates can directly use alloy_primitives::Log, potentially improving readability and maintainability.

Cons:
Tight Coupling: Changes to alloy_primitives::Log might impact all crates using it, requiring careful coordination and versioning.
Limited Flexibility: Adding methods specific to reth or revm requires modifying the shared type, potentially leading to conflicts or unnecessary features in other crates.


Our current usage on primitives type is nearly identical, so I think going with Option 2 is good.

**Comment by mattsse** [comment]
ty for this assessment.

given we can add/change the alloy types since we maintain it, option 2 is he best and I'd like to proceed with this.

The same applies to `alloy_eips::AccessList`

**Comment by 0xqd** [comment]
one test vector added, based from the output of proptest current `reth_primitives::log` encoded, and decoded.

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, we want a few more cases, and they should be dumped as json files in a separate folder

**Comment by 0xqd** [comment]
@mattsse done, can you help checking if we miss anything else.

**Review by mattsse — CHANGES_REQUESTED** [review]
actually, I have a better idea how to test this,

what we can do, is write a test in reth-primitives that generates arbitrary reth-primitives::Log and uses those for roundtrip tests against alloy-primitives::Log
then once we do replace the reth-primitives::Log type, we simply move the type into the test module

**Comment by 0xqd** [comment]
That's exactly what I did :D 
https://github.com/0xqd/reth/commit/a770392c20400ff0693db319e6ad6e46e9ed5e6f

Then the output being add to a vector to be read again later by current `reth_primitives` `alloy/log.rs`.

File: crates/storage/codecs/src/alloy/log.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
```
Comment (mattsse) [inline crates/storage/codecs/src/alloy/log.rs:12]: here we're allocating a new vec
we can use `split()` here

File: crates/storage/codecs/testdata/log_compact.json
Comment (mattsse) [inline crates/storage/codecs/testdata/log_compact.json:1]: can we change the fields from byte arrays to hex encoded strings? and minify?

**Review by mattsse — CHANGES_REQUESTED** [review]
> That's exactly what I did :D

cool, can you add this as a test in reth-primitives?

**Review by mattsse — APPROVED** [review]
lgtm

once merged we can replace reth-primitives::Log with alloy-primitives::log by exporting it from primitives and update usage

pending @joshieDo 

here's to code on main, Logflags is a zst so we can get rid of it

```rust

impl Compact for Log {
        fn to_compact<B>(self, buf: &mut B) -> usize
        where
            B: bytes::BufMut + AsMut<[u8]>,
        {
            let mut flags = LogFlags::default();
            let mut total_length = 0;
            let mut buffer = bytes::BytesMut::new();
            let address_len = self.address.to_compact(&mut buffer);
            let topics_len = self.topics.specialized_to_compact(&mut buffer);
            let data_len = self.data.to_compact(&mut buffer);
            let flags = flags.into_bytes();
            total_length += flags.len() + buffer.len();
            buf.put_slice(&flags);
            buf.put(buffer);
            total_length
        }
        fn from_compact(mut buf: &[u8], len: usize) -> (Self, &[u8]) {
            let (flags, mut buf) = LogFlags::from(buf);
            let (address, new_buf) = Address::from_compact(buf, buf.len());
            buf = new_buf;
            let (topics, new_buf) = Vec::specialized_from_compact(buf, buf.len());
            buf = new_buf;
            let mut data = Bytes::new();
            (data, buf) = Bytes::from_compact(buf, buf.len() as usize);
            let obj = Log {
                address: address,
                topics: topics,
                data: data,
            };
            (obj, buf)
        }
    }
```

**Review by joshieDo — APPROVED** [review]

**Pull request merged**